sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java

```diff
@@ -394,6 +418,7 @@ public void startBundle(StartBundleContext c) {
 
     @ProcessElement
     public void processElement(ProcessContext c, BoundedWindow window) throws Exception {
+      sink.getDynamicDestinations().setSideInputAccessorFromProcessContext(c);
```

Comment:
This pattern is brittle. You should leave `DynamicDestinations` immutable and pass this explicitly to methods that need it.

Example patterns that might make sense:

```
sink.getDynamicDestinations(new SideInputAccessor(processContext))

sink.getDynamicDestinations()
    .getDestination(new SideInputAccessor(processContext))
```